File: PermuteSum/PermuteSum.java
// PermuteSum.java
import java.util.*;

public class PermuteSum {
    public static void main(String[] args){
	Set<String> values = new HashSet<String>();
	int targetSum = 5;
	int permuteLength = 3;
	int[] permuteSet = {0, 1, 2, 3};

	generatePermuteSum(permuteSet, permuteLength, targetSum, values, "", permuteLength);
	System.out.println(values);
    }

    public static boolean generatePermuteSum(int[] set, int length, int sum, Set<String> values, String x, int originalLength){
    	if (x.length() == originalLength && sum(x) == sum){ values.add(x); }
	for (int i = 0; i < set.length && length > 0; i++){
	    generatePermuteSum(set, length - 1, sum, values, x + Integer.toString(set[i]), originalLength);
	}

	return values.isEmpty();

    }

    public static int sum(String s){
	int sumValue = 0;
	for (int i = 0; i < s.length(); i++){
	    sumValue += s.charAt(i) - '0';
	}

	return sumValue;
    }
}

File: PermuteSum/PermuteSum.py
## PermuteSum.py
def main():
    values = set()
    length = 3
    sum_value = 5
    value_set = [0, 1, 2, 3]
    empty = []
    permute_sum(value_set, length, sum_value, values, empty, length)

def permute_sum(value_set, length, sum_value, values, x, original_length):
    if (len(x) == original_length and sum(x) == sum_value):
        values.add(x.__str__())

    i = 0
    while i < len(value_set) and length > 0:
        y = list(x) + [value_set[i]]
        permute_sum(value_set, length - 1, sum_value, values, y, original_length)
        i += 1

main()

